daemon/engines/deezer_engine.py

```diff
@@ -0,0 +1,84 @@
+from dataclasses import dataclass
+from engine import Engine, SongChoice
+
+import re
+import deemix
+import deemix.settings
+from deemix.downloader import Downloader
+from deezer import Deezer, TrackFormats
+from deezer.errors import DataException
+
+class DeezerEngine(Engine):
+    name = "Deezer"
+    description = "Plays from deezer"
+    # Deezer has high-quality audio and great tagging
+    rank = 8
+
+    def __init__(self, library, options):
+        if not 'arl' in options:
+            raise Exception('No arl given to deezer engine, but required to function')
+        self.dz = Deezer()
+        self.dz.login_via_arl(options['arl'])
+        self.library = library
+
+    @dataclass
+    class DeezerSongChoice(SongChoice):
+        dz: Deezer
+        link: str
+        dir: str
+
+        def choose(self):
+            downloader = deemix.generateDownloadObject(
+                self.dz,
+                self.link,
+                TrackFormats.MP3_320
+            )
+            self.dl = Downloader(self.dz, downloader, {
+                **deemix.settings.DEFAULTS,
+                'downloadLocation': self.dir,
+            }, listener = self).start()
+            return self.path
+
+        def send(self, kind, message):
+            if kind != 'updateQueue':
+                return
+            if 'downloaded' in message and message['downloaded']:
+                self.path = message['downloadPath']
+
+    def search(self, query: str):
+        url_check = query
+        if re.search("([a-z]*[.])\w+", url_check):
+            return []
+        results = self.dz.api.search(query)['data']
+        res_list = [
+            self.DeezerSongChoice(
+                title = x['title'],
+                artist = x['artist']['name'],
+                dz = self.dz,
+                link = x['link'],
+                dir = self.library.path()
+            )
+            for x in results[0:20]
+        ]
+        # If there are no search results, try an id-search if the query can
+        # be an int.
+        if not res_list and int(query):
```

Comment:
If the query reaches here and is not an int, this will throw an exception because of the `int()` conversion. Intended?

Reply:
Not added in this PR as far as i know. I already opened an issue for it #66, but it should be done separately as this merge is already quite big ^^

Reply:
I just fixed it now anyway :P